File: src/main/java/com/primogemstudio/advancedfmk/network/UpdatePacket.java
package com.primogemstudio.advancedfmk.network;

import com.primogemstudio.advancedfmk.mmd.entity.TestEntity;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientEntityEvents;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.PacketSender;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientPacketListener;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;

import static com.primogemstudio.advancedfmk.AdvancedFramework.MOD_ID;

public class UpdatePacket implements ClientPlayNetworking.PlayChannelHandler {
    private static final Int2ObjectMap<String> map = new Int2ObjectOpenHashMap<>();

    static {
        ClientEntityEvents.ENTITY_LOAD.register((entity, world) -> {
            if (entity instanceof TestEntity te) {
                if (map.containsKey(te.getId())) {
                    te.setModel(map.get(te.getId()));
                    map.remove(te.getId());
                }
            }
        });
    }

    public static void flush() {
        var level = Minecraft.getInstance().level;
        if (level == null) return;
        for (var it : map.int2ObjectEntrySet()) {
            if (level.getEntity(it.getIntKey()) instanceof TestEntity te) {
                te.setModel(it.getValue());
            }
        }
        map.clear();
    }

    public static void send(TestEntity entity, String path) {
        var buf = PacketByteBufs.create();
        buf.writeInt(entity.getId());
        buf.writeUtf(path);
        entity.level().players().forEach(p -> ServerPlayNetworking.send((ServerPlayer) p, new ResourceLocation(MOD_ID, "update"), buf));
    }

    @Override
    public void receive(Minecraft client, ClientPacketListener handler, FriendlyByteBuf buf, PacketSender responseSender) {
        var id = buf.readInt();
        var path = buf.readUtf();
        map.put(id, path);
    }
}

File: src/main/java/com/primogemstudio/advancedfmk/network/UpdatePacket.kt
package com.primogemstudio.advancedfmk.network

import com.primogemstudio.advancedfmk.client.MOD_ID
import com.primogemstudio.advancedfmk.mmd.entity.TestEntity
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientEntityEvents
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.client.Minecraft
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player

class UpdatePacket(private val id: Int, private val path: String) : CustomPacketPayload {
    companion object {
        private val map = Int2ObjectOpenHashMap<String>()
        val TYPE = CustomPacketPayload.Type<UpdatePacket>(ResourceLocation(MOD_ID, "update"))
        val CODEC = StreamCodec.ofMember(UpdatePacket::write, ::UpdatePacket)!!

        init {
            ClientEntityEvents.ENTITY_LOAD.register { ent, _ ->
                if (ent is TestEntity && map.containsKey(ent.id)) {
                    ent.setModel(map[ent.id])
                    map.remove(ent.id)
                }
            }
        }

        fun flush() {
            val level = Minecraft.getInstance().level ?: return
            map.int2ObjectEntrySet().forEach {
                with(level.getEntity(it.intKey)) {
                    if (this is TestEntity) setModel(it.value)
                }
            }
            map.clear()
        }

        fun send(entity: TestEntity, path: String?) {
            entity.level().players().forEach { p: Player? ->
                ServerPlayNetworking.send(
                    p as ServerPlayer?, UpdatePacket(entity.id, path ?: "")
                )
            }
        }
    }

    fun write(buf: FriendlyByteBuf) {
        buf.writeInt(id)
        buf.writeUtf(path)
    }

    constructor(buf: FriendlyByteBuf) : this(buf.readInt(), buf.readUtf())


    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
        return TYPE
    }

    object Handler : ClientPlayNetworking.PlayPayloadHandler<UpdatePacket> {
        override fun receive(pk: UpdatePacket, context: ClientPlayNetworking.Context) {
            map[pk.id] = pk.path
        }
    }
}